crates/biome_js_analyze/src/lint/nursery/no_irregular_whitespace.rs

```diff
@@ -0,0 +1,103 @@
+use biome_analyze::{context::RuleContext, declare_lint_rule, Ast, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::AnyJsStatement;
+use biome_rowan::{AstNode, Direction};
+
+const IRREGULAR_WHITESPACES: &[char; 22] = &[
+    '\u{c}', '\u{b}', '\u{85}', '\u{feff}', '\u{a0}', '\u{1680}', '\u{180e}', '\u{2000}',
+    '\u{2001}', '\u{2002}', '\u{2003}', '\u{2004}', '\u{2005}', '\u{2006}', '\u{2007}', '\u{2008}',
+    '\u{2009}', '\u{200a}', '\u{200b}', '\u{202f}', '\u{205f}', '\u{3000}',
+];
+
+declare_lint_rule! {
+    /// Disallows the use of irregular whitespace characters.
+    ///
+    /// Invalid or irregular whitespace causes issues with ECMAScript 5 parsers and also makes code harder to debug.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// constcount=1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = 'thing';
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const count = 1;
+    /// ```
+    ///
+    /// ```js
+    /// const foo = '';
+    /// ```
+    ///
+    pub NoIrregularWhitespace {
+        version: "next",
+        name: "noIrregularWhitespace",
+        language: "js",
+        recommended: false,
+    }
+}
+
+impl Rule for NoIrregularWhitespace {
+    type Query = Ast<AnyJsStatement>;
```

Comment:
Here's the thing about this query, and it doesn't play well with the logic of the rule. This returns any statement, and in this example, you have[ multiple statements](https://biomejs.dev/playground/?code=YwBvAG4AcwB0ACAAZgAgAD0AIABmAHUAbgBjAHQAaQBvAG4AIABmACgAKQAgAHsACgAJAGkAZgAgACgAdAByAHUAZQApACAAewAKAAoACQB9ACAAZQBsAHMAZQAgAHsACgAKAAkAfQAKAH0A):

```js
const f = function f() {
	if (true) {} else {}
}
```

The query will trigger for all the statements of this snippet. However, for each statement, you retrieve ALL tokens, which means that, for example, the `else` token is checked **multiple** times because it is returned by all the statements you're querying.

That's why you have this massive regression in performance; the rule is doing repeated work, A LOT. I believe the best strategy is to query the **root** node of the CST, so you're sure that you aren't doing repeated work on the same tokens.

Reply:
I see, I can make the change to query only the root node (I assume it's the `JsModule` module one?), but that would mean that we can only have one error report about irregular whitespaces per file, right? As far as I know there's no way to report multiple error ranges on a lint rule

Reply:
> I see, I can make the change to query only the root node (I assume it's the `JsModule` module one?), but that would mean that we can only have one error report about irregular whitespaces per file, right? As far as I know there's no way to report multiple error ranges on a lint rule

Yes, there is: https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#multiple-signals

Reply:
oooh, okay, I'm gonna dig into that, thanks!

Reply:
just did this change, worked like a charm

Reply:
nevermind, the benchmarks still failed, I will keep working on that